genrl/core/actor_critic.py

```diff
@@ -40,14 +42,121 @@ def __init__(
+    """
+
+    def __init__(
+        self,
+        state_dim: spaces.Space,
+        action_dim: spaces.Space,
+        shared_layers: Tuple = (32, 32),
+        policy_layers: Tuple = (32, 32),
+        value_layers: Tuple = (32, 32),
+        val_type: str = "V",
+        discrete: bool = True,
+        **kwargs,
+    ):
+        super(MlpSharedActorCritic, self).__init__()
+        self.shared = mlp([state_dim] + list(shared_layers))
+        self.actor = MlpPolicy(
+            shared_layers[-1], action_dim, policy_layers, discrete, **kwargs
+        )
+        self.critic = MlpValue(
+            shared_layers[-1], action_dim, val_type, value_layers, **kwargs
+        )
+        self.state_dim = state_dim
+        self.action_dim = action_dim
+
+    def get_params(self):
+        actor_params = list(self.shared.parameters()) + list(self.actor.parameters())
+        critic_params = list(self.shared.parameters()) + list(self.critic.parameters())
+        return actor_params, critic_params
+
+    def get_features(self, state: torch.Tensor):
+        """Extract features from the state, which is then an input to get_action and get_value
+
+        Args:
+            state (:obj:`torch.Tensor`): The state(s) being passed
+
+        Returns:
+            features (:obj:`torch.Tensor`): The feature(s) extracted from the state
+        """
+        features = self.shared(state)
+        return features
+
+    def get_action(self, state: torch.Tensor, deterministic: bool = False):
+        """Get Actions from the actor
+
+        Arg:
+            state (:obj:`torch.Tensor`): The state(s) being passed to the critics
+            deterministic (bool): True if the action space is deterministic, else False
+
+        Returns:
+            action (:obj:`list`): List of actions as estimated by the critic
+            distribution (): The distribution from which the action was sampled
+                            (None if determinist
+        """
+
+        state = torch.as_tensor(state).float()
+        features = self.get_features(state)
+        action_probs = self.actor(features)
+        action_probs = nn.Softmax(dim=-1)(action_probs)
+
+        if deterministic:
+            action = torch.argmax(action_probs, dim=-1).unsqueeze(-1).float()
+            distribution = None
+        else:
+            distribution = Categorical(probs=action_probs)
+            action = distribution.sample()
+
+        return action, distribution
+
+    def get_value(self, state: torch.Tensor):
+        """Get Values from the Critic
+
+        Arg:
+            state (:obj:`torch.Tensor`): The state(s) being passed to the critics
+
+        Returns:
+            values (:obj:`list`): List of values as estimated by the critic
+        """
+        state = torch.as_tensor(state).float()
+        if self.critic.val_type == "Qsa":
+            features = self.shared(state[:, :, :-1])
+            features = torch.cat([features, state[:, :, -1].unsqueeze(-1)], dim=-1)
```

Comment:
What is this indexing over here? Are you sure you're not missing any corner cases here?